nav2-usability/scenario-code/nested-routing/router.dart

```diff
@@ -0,0 +1,327 @@
+          if (idx == 0) {
+            _appState.page = AppPage.newBooks;
+          } else {
+            _appState.page = AppPage.settings;
+          }
+        },
+        items: [
+          BottomNavigationBarItem(
+            label: 'Books',
+            icon: Icon(Icons.chrome_reader_mode_outlined),
+          ),
+          BottomNavigationBarItem(
+            label: 'Settings',
+            icon: Icon(Icons.settings),
+          ),
+        ],
+      ),
+    );
+  }
+}
+
+class AppRoutePath {}
+
+class NewBooksRoutePath extends AppRoutePath {}
+
+class AllBooksRoutePath extends AppRoutePath {}
+
+class SettingsRoutePath extends AppRoutePath {}
+
+class BooksScreen extends StatefulWidget {
+  final AppPage appPage;
+  final ValueChanged<int> onTabSelected;
+
+  BooksScreen({
+    Key? key,
+    required this.appPage,
+    required this.onTabSelected,
+  }) : super(key: key);
+
+  @override
+  _BooksScreenState createState() => _BooksScreenState();
+}
+
+class _BooksScreenState extends State<BooksScreen>
+    with SingleTickerProviderStateMixin {
+  late final TabController _tabController;
+
+  @override
+  void initState() {
+    super.initState();
+    var initialIndex = widget.appPage == AppPage.newBooks ? 0 : 1;
+
+    _tabController =
+        TabController(length: 2, vsync: this, initialIndex: initialIndex);
+  }
+
+  @override
+  void didUpdateWidget(BooksScreen oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.appPage == AppPage.newBooks) {
+      _tabController.index = 0;
+    } else {
+      _tabController.index = 1;
+    }
+  }
+
+  void dispose() {
+    _tabController.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Column(
+      children: [
+        TabBar(
+          controller: _tabController,
+          onTap: widget.onTabSelected,
+          labelColor: Theme.of(context).primaryColor,
+          tabs: [
+            Tab(icon: Icon(Icons.bathtub), text: 'New'),
+            Tab(icon: Icon(Icons.group), text: 'All'),
+          ],
+        ),
+        Expanded(
+          child: TabBarView(
+            controller: _tabController,
+            children: [
+              NewBooksScreen(),
+              AllBooksScreen(),
+            ],
+          ),
+        ),
+      ],
+    );
+  }
+}
+
+class SettingsScreen extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      body: Center(
+        child: Text('Settings'),
+      ),
+    );
+  }
+}
+
+class AllBooksScreen extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      body: Center(
+        child: Text('All Books'),
+      ),
+    );
+  }
+}
+
+class NewBooksScreen extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      body: Center(
+        child: Text('New Books'),
+      ),
+    );
+  }
+}
+
+class FadeTransitionPage extends Page {
```

Comment:
Since you are mutating the route content by the page.child. You would need to write this page class in this way. Ideally we should make this easier by adding a convenient page class for PageRouteBuilder. I am still thinking how we can do that because it may cause some scope issue.
```dart
class FadeTransitionPage extends Page {
  final Widget child;

  FadeTransitionPage({LocalKey key, @required this.child}) : super(key: key);

  @override
  Route createRoute(BuildContext context) {
    return PageBasedFadeTransitionRoute(this);
  }
}
class PageBasedFadeTransitionRoute extends PageRoute {
  PageBasedFadeTransitionRoute(this.page) : super(
    settings: page,
  );

  final FadeTransitionPage page;

  @override
  Color get barrierColor => null;

  @override
  String get barrierLabel => null;

  @override
  Duration get transitionDuration => const Duration(milliseconds: 300);

  @override
  bool get maintainState => true;

  @override
  Widget buildPage(BuildContext context, Animation<double> animation, Animation<double> secondaryAnimation) {
    var curveTween = CurveTween(curve: Curves.easeIn);
    return FadeTransition(
      opacity: animation.drive(curveTween),
      child: (settings as FadeTransitionPage).child,
    );
  }

  @override
  Widget buildTransitions(BuildContext context, Animation<double> animation, Animation<double> secondaryAnimation, Widget child) {
    return child;
  }
}
```

Reply:
I filed a issue to track it https://github.com/flutter/flutter/issues/80804